pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/InvertedIndexAndDictionaryBasedForwardIndexCreator.java

```diff
@@ -0,0 +1,609 @@
+                    + "column %s", maxNumberOfMultiValues[0], _maxNumberOfMultiValues, _columnName));
+      }
+
+      // Calculate value index for each docId in the forward index value buffer
+      // Re-use forward index length buffer to store the value index for each docId, where value index is the index in
+      // the forward index value buffer where we should put next dictId for the docId
+      int forwardValueIndex = 0;
+      for (int docId = 0; docId < _numDocs; docId++) {
+        int length = getInt(_forwardIndexLengthBuffer, docId);
+        putInt(_forwardIndexLengthBuffer, docId, forwardValueIndex);
+        forwardValueIndex += length;
+      }
+
+      // Construct the forward index values buffer from the inverted index using the length buffer for index tracking
+      for (int dictId = 0; dictId < _cardinality; dictId++) {
+        ImmutableRoaringBitmap docIdsBitmap = invertedIndexReader.getDocIds(dictId);
+        int finalDictId = dictId;
+        docIdsBitmap.stream().forEach(docId -> {
+          int index = getInt(_forwardIndexLengthBuffer, docId);
+          putInt(_forwardIndexValueBuffer, index, finalDictId);
+          putInt(_forwardIndexLengthBuffer, docId, index + 1);
+          if (shouldTrackVarLengthMetadata) {
+            trackMaxRowLengthInBytes(dictionary, maxRowLengthInBytes, docId, finalDictId);
+          }
+        });
+      }
+
+      IndexCreationContext.Forward context =
+          IndexCreationContext.builder().withIndexDir(_segmentMetadata.getIndexDir())
+              .withColumnMetadata(_columnMetadata).withforwardIndexDisabled(false).withDictionary(_dictionaryEnabled)
+              .withTotalNumberOfEntries(_nextValueId).withMaxNumberOfMultiValueElements(maxNumberOfMultiValues[0])
+              .withMaxRowLengthInBytes(maxRowLengthInBytes[0]).withLengthOfLongestEntry(lengthOfLongestEntry[0])
+              .build().forForwardIndex(_chunkCompressionType, _indexLoadingConfig.getColumnProperties());
+
+      writeToForwardIndex(dictionary, context);
+
+      // Setup and return the metadata properties to update
+      Map<String, String> metadataProperties = new HashMap<>();
+      metadataProperties.put(getKeyFor(_columnName, HAS_DICTIONARY), String.valueOf(_dictionaryEnabled));
+      metadataProperties.put(getKeyFor(_columnName, DICTIONARY_ELEMENT_SIZE),
+          String.valueOf(_dictionaryEnabled ? _columnMetadata.getColumnMaxLength() : 0));
+      metadataProperties.put(getKeyFor(_columnName, MAX_MULTI_VALUE_ELEMENTS),
+          String.valueOf(maxNumberOfMultiValues[0]));
+      metadataProperties.put(getKeyFor(_columnName, TOTAL_NUMBER_OF_ENTRIES),
+          String.valueOf(_nextValueId));
+      return metadataProperties;
+    }
+  }
+
+  private void trackLengthOfLongestEntry(Dictionary dictionary, int[] lengthOfLongestEntry, int dictId) {
+    switch (_columnMetadata.getFieldSpec().getDataType().getStoredType()) {
```

Comment:
(nit) this can be passed in as argument or make it a local variable to make the code more clean